Source hugo (#2)
LucasRoesler/lucasroesler.github.io: content/posts/2021/test-python-functions/index.md

---
title: "Using Tox and PyTest with OpenFaaS"
description: "How to Unit Test Python functions in OpenFaaS"
date: 2021-03-06T00:00:00+02:00
tags:
  - programming
  - openfaas
  - serverless
  - python
  - pytest
  - tox
  - ci/cd
  - github actions
  - testing
  - unit test
draft: false
# <span>Photo by <a href="https://unsplash.com/@alex_andrews?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText">Alexander Andrews</a> on <a href="https://unsplash.com/?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText">Unsplash</a></span>
coverImage: images/cover2-alexander-andrews.jpg
---

> I think it is an uncontroversial statement that to say testing is important in software development. Writing tests may not always be fun, but nothing is a sweet as that moment when a unit test catches a bug before your commit.

Reviewer: ```suggestion
I think it is an uncontroversial statement to say testing is important in software development. Writing tests may not always be fun, but nothing is as sweet as that moment when a unit test catches a bug before your commit.
```

Reviewer: Also wondering here.. will the unit test catch a bug before your COMMIT?  Or before your deployment?

> In OpenFaaS we have tons of tests in each project, even the [`certifier`](https://github.com/openfaas/certifier) itself that runs a short suite of end-to-end tests. _But_, not all of our function templates have first class testing support. For Go, Node, and Java, this has already been added. For Go it is easy because there is a standard test command `go test`. For Node and Java we (or the community) have picked and optional default test command. Soon, we are going to do the same for Python.

Reviewer: the second half of the first sentence makes sense to me.  Should it be `even the certifier itself runs a short suite of end-to-end tests`?

But what do you do in the mean time? Or, what if you have your own Python template and you want to add testing to it?  In this post I am going to show you how to use  [`tox`](https://tox.readthedocs.io/en/latest/) and [`pytest`](https://docs.pytest.org/en/stable/) with any Python OpenFaaS template.

<!-- more -->
## Intro

For this demo we are going to create a tiny calculator function. This gives us just enough code to demonstrate interesting but very common test cases:

> 1. is the request correct, .e.g did the request specify a valid math operation and two numbers,

Reviewer: I think this is also `i.e.` because you're explaining it better?  But I could be wrong.

2. is the response correct, i.e. the calculation.

Almost all functions will want these kind of tests and, as we will see, they are easy to write.

If you want to skip to the final result, checkout the sample repo here: https://github.com/LucasRoesler/pytest-openfaas-sample

### Start the project
```sh
$ mkdir pytest-sample
$ cd pytest-sanmple
$ faas-cli template store pull python3-flask
$ mv calc.yml stack.yml
```

**Note** Don't forget to add this section to your stack file so that the `faas-cli` will know how to automatically pull the template

```yaml
configuration:
  templates:
    - name: python3-flask
      source: https://github.com/openfaas/python-flask-template

```

### Setup the local python environment
I use [conda](https://docs.conda.io/projects/conda/en/latest/) for my local virtual environments, but you can of course use [`virtualenv`](https://virtualenv.readthedocs.org/en/latest/) or [`venv`](https://docs.python.org/3/library/venv.html) (see also [RealPython's primer on virtual environments](https://realpython.com/python-virtual-environments-a-primer/)).

> In short this creates an isolated and repeatable development environment which you can delete and recreate if you need.

Reviewer: Cut the `In short`

```sh
$ conda create -n pytest-sample tox
$ conda activate pytest-sample
$ cat <<EOF >> requirements.txt
pydantic==1.7.3
flask==1.1.2
EOF
$ cat <<EOF >> dev.txt
tox
pytest
black
pylint
EOF
$ conda install --yes --file requirements.txt
$ conda install --yes --file dev.txt
```

Now we are ready to develop the calculator.

### Add the first tests
Let's start with some simple tests:

```sh
$ touch calc/handler_test.py
```

then add the following test cases for a couple of our happy paths

```py
import calc.handler as h

class TestParsing:
    def test_operation_addition(self):
        req = '{"op": "+", "var1": "1.0", "var2": 0}'
        resp, code = h.handle(req)
        assert code == 200
        assert resp["value"] == 1.0

    def test_operation_multiplication(self):
        req = '{"op": "*", "var1": "100.01", "var2": 1}'
        resp, code = h.handle(req)
        assert code == 200
        assert resp["value"] == 100.01
```

> At this point, we have followed normal conventions for pytest, by default, pytest will look for files that match the pattern `*_test.py` and then the run the functions that  match `test_*` (including matching methods if the class is named `Test*`, which is what we did above).

Reviewer: either `as we have followed normal conventions for pytest, by default pytest will look` (removing the comma after the default)

or `we have followed normal conventions for pytest, so pytest will by default`

> Now, we can run our test and see all of the errors in their wonderful glory

Reviewer: ```suggestion
Now, we can run our test and see all of the errors in their wonderful glory.
```

self = <calc.handler_test.TestParsing object at 0x7fc8029ad820>

    def test_operation_addition(self):
        req = '{"op": "+", "var1": "1.0", "var2": 0}'
>       resp, code = h.handle(req)
E       ValueError: too many values to unpack (expected 2)

calc/handler_test.py:49: ValueError
______________________________ TestParsing.test_operation_multiplication ______________________________

self = <calc.handler_test.TestParsing object at 0x7fc8029add30>

    def test_operation_multiplication(self):
        req = '{"op": "^", "var1": "2", "var2": -2}'
>       resp, code = h.handle(req)
E       ValueError: too many values to unpack (expected 2)

calc/handler_test.py:73: ValueError
======================================= short test summary info =======================================
FAILED calc/handler_test.py::TestParsing::test_operation_addition - ValueError: too many values to u...
FAILED calc/handler_test.py::TestParsing::test_operation_multiplication - ValueError: too many value...
========================================== 2 failed in 0.05s ==========================================
```

> In this case we get a value error because our handler function only returns a string, not a dictionary and a status code like the test expects.  These test cases  are expecting that the handler returns something that looks like this

Reviewer: ```suggestion
In this case we get a value error because our handler function only returns a string, not a dictionary and a status code like the test expects.  These test cases expect the handler to return something that looks like this
```

```py
return {"value": 1.1}, 200
```

> This return value works with Flask, it will json serialize the dictionary and set the status code to 200 for us. Like the above snippet and run pytest again, the we get a new error, a value error because we got the wrong calculation result

Reviewer: ```suggestion
This return value works with Flask; it will json serialize the dictionary and set the status code to 200 for us. Like the above snippet and run pytest again, then we get a new error, a value error because we got the wrong calculation result
```

Reviewer: I have no idea what the second sentence is trying to say.  Are you sure you're trying to say `Like the above snippet` ?

platform linux -- Python 3.8.8, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /home/lucas/code/openfaas/sandbox/pytest-sample
collected 2 items

calc/handler_test.py .F                                                                         [100%]

============================================== FAILURES ===============================================
______________________________ TestParsing.test_operation_multiplication ______________________________

self = <calc.handler_test.TestParsing object at 0x7f9f30ecbd30>

    def test_operation_multiplication(self):
        req = '{"op": "^", "var1": "2", "var2": -2}'
        resp, code = h.handle(req)
        assert code == 200
>       assert resp["value"] == 0.25
E       assert 1 == 0.25

calc/handler_test.py:75: AssertionError
======================================= short test summary info =======================================
FAILED calc/handler_test.py::TestParsing::test_operation_multiplication - assert 1 == 0.25
===================================== 1 failed, 1 passed in 0.06s =====================================
```

### Adding `tox`

> Before we fix the test, we want to setup one more thing: [`tox`](https://tox.readthedocs.io/en/latest/index.html): `tox`provides a unified set of tooling that will run `pytest` for us. This provides a standardized interface that we will use later in CI. In fact, it ensures we are running exactly the same thing in CI and our local dev environment, hopefully reducing the numer of "Actually fix CI" commit messages.  Later, if you decide that you prefer `nose` or some other testing tool or if you application is even more complex and requires pre-test configuration -- we can manage that through `tox`.

Reviewer: ```suggestion
Before we fix the test, we want to setup one more thing: [`tox`](https://tox.readthedocs.io/en/latest/index.html): `tox`provides a unified set of tooling that will run `pytest` for us. This provides a standardized interface that we will use later in CI. In fact, it ensures we are running exactly the same thing in CI as our local dev environment, hopefully reducing the number of "Actually fix CI" commit messages.  Later, if you decide that you prefer `nose` or some other testing tool or if you application is even more complex and requires pre-test configuration -- we can manage that through `tox`.
```
